app/scripts/helpers/GeometryWorker.js

```diff
@@ -0,0 +1,189 @@
+            var part = geometries[i].vertices.slice(verticesPossition, verticesPossition + chunkSize);
+            verticesPossition += chunkSize;
+            self.postMessage({msg: 'vertices', geometries: part, nodeId: nodeId});
+        }
+        else if(geometries[i].linestrip && geometries[i].linestrip.length > linestripPossition) {
+            var part = geometries[i].linestrip.slice(linestripPossition, linestripPossition + chunkSize);
+            linestripPossition += chunkSize;
+            self.postMessage({msg: 'linestrip', geometries: part, nodeId: nodeId});
+        }
+        else if(geometries[i].faces && geometries[i].faces.length > facesPossition) {
+            var part = geometries[i].faces.slice(facesPossition, facesPossition + chunkSize / 2);
+            facesPossition += chunkSize / 2;
+            self.postMessage({msg: 'faces', geometries: part, nodeId: nodeId});
+        } else {
+            i++;
+            verticesPossition = 0;
+            linestripPossition = 0;
+            facesPossition = 0;
+
+            if(geometries.length > i)
+                self.postMessage({msg: 'NewPart', i : i, nodeId: nodeId});
+        }
+
+        sleep(50);
+    }
+};
+
+sleep =function (ms) 
+{
+    var e = new Date().getTime() + (ms);
+    while (new Date().getTime() <= e) {}
+};
+
+addPoints = function (graphicData, geometries) {
+    var length, pos;
+    // if we have single points
+    if (graphicData.pointVertices && graphicData.pointVertices.length) {
+        length = graphicData.pointVertices.length / graphicData.numberOfCoordinates;
+        var points = { vertices: [] }, onePoint;
+        for (var i = 0; i < length; i++) {
+            // add [x, y, z]
+            pos = i * graphicData.numberOfCoordinates;
+            onePoint = [
+                graphicData.pointVertices[pos],
+                graphicData.pointVertices[pos + 1],
+                graphicData.pointVertices[pos + 2]
+            ];
+            points.vertices.push(onePoint);
+        }
+
+        geometries.push(points);
+    }
+};
+
+addCurves = function (graphicData, geometries) {
+    // if we have line strips
+    if (graphicData.lineStripVertices && graphicData.lineStripVertices.length
+        && graphicData.lineStripCounts && graphicData.lineStripCounts.length) {
+        var curve, count, oneVertex, size, pos, i, index = 0;
+
+        for (var k = 0; k < graphicData.lineStripCounts.length; k++) {
+            curve = { linestrip: [] };
+            count = graphicData.lineStripCounts[k];
+
+            if (!count) {
+                continue;
+            }
+
+            size = index + count;
+
+            for (i = index; i < size; i++) {
+                pos = i * graphicData.numberOfCoordinates;
+                oneVertex = [
+                    graphicData.lineStripVertices[pos],
+                    graphicData.lineStripVertices[pos + 1],
+                    graphicData.lineStripVertices[pos + 2]
+                ];
```

Comment:
Hi @pbidenko,

It occurs to me that "unpacking" the vertices in this way is probably completely unnecessary and very expensive.  Think about how many new arrays we have to allocate in order to do it.  

Why not just keep everything in one big horizontal array?  This would require some modifications to ThreeCSG, but would probably make this whole thing far faster.  Let me know what you think.  I know I've thought of this rather late in the game.  Better late than never, I suppose.